### PR TITLE
DDF-1950 CswSource Refactoring

### PR DESCRIPTION
**Pull request by clockard**
#### What does this PR do?
Refactors the CswSource to and AbstractCswSource in a new module for future extension.
#### Who is reviewing it?
@kcwire @tbatie @gordocanchola @vinamartin 
#### How should this be tested?
Need to make sure all unit and itests pass
#### Any background context you want to provide?
Registry and other programs will be extending the new AbstractCswSource
Another follow on ticket will be responsible for further breaking down the AbstractCswSource to be more manageable and extensible 
#### What are the relevant tickets?
DDF-1950
#### Checklist:
- [ ] Documentation Updated
- [X] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

- Moved CswSource to AbstractCswSource and added abstract method getAdditionalConsumers
- Move all of the current csw source classes to the spacial-csw-common package
- Created a new CswSourceImpl that implements the new AbstractCswSource

### CODE REVIEW

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReader.java [hunk 1]
```diff
@@ -35,7 +35,7 @@
 import org.codice.ddf.spatial.ogc.csw.catalog.common.CswConstants;
 import org.codice.ddf.spatial.ogc.csw.catalog.common.CswRecordCollection;
 import org.codice.ddf.spatial.ogc.csw.catalog.common.CswSourceConfiguration;
-import org.codice.ddf.spatial.ogc.csw.catalog.converter.GetRecordsResponseConverter;
+//import org.codice.ddf.spatial.ogc.csw.catalog.converter.GetRecordsResponseConverter;
```
Comment (vinamartin) [inline catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReader.java]: Remove

File: catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java [hunk 2]
```diff
@@ -214,7 +185,7 @@ public void testGetMultipleMetacardsWithForeignText() throws Exception {
                 is);
         List<Metacard> metacards = cswRecords.getCswRecords();
         assertThat(metacards, not(nullValue()));
-        assertThat(metacards.size(), equalTo(10));
+        assertThat(metacards.size(), equalTo(1));
     }
 
     // //////////////////////////////////////////////////////////////////////////////////////////////////////
```
Comment (vinamartin) [inline catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java]: What is this thing anyway? I don't think you touched it, but it should probably be cleaned up.

File: catalog/spatial/csw/spatial-csw-source/pom.xml [hunk 3]
```diff
@@ -207,22 +213,22 @@
                                         <limit>
                                             <counter>INSTRUCTION</counter>
                                             <value>COVEREDRATIO</value>
-                                            <minimum>0.62</minimum>
+                                            <minimum>0.0</minimum>
```
Comment (vinamartin) [inline catalog/spatial/csw/spatial-csw-source/pom.xml]: What happened to the coverage here? Does it have to do with making it abstract?

File: catalog/spatial/csw/spatial-csw-source-common/pom.xml [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.jvnet.ogc</groupId>
+            <artifactId>ogc-tools-gml-jts</artifactId>
+            <version>1.0.3</version>
+        </dependency>
+        <dependency>
+            <groupId>xmlunit</groupId>
+            <artifactId>xmlunit</artifactId>
+            <version>1.6</version>
```
Comment (tbatie) [inline catalog/spatial/csw/spatial-csw-source-common/pom.xml]: We are using 1.4 everywhere else in ddf, is there a reason you need 1.6 here?

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReader.java [hunk 5]
```diff
@@ -63,11 +63,10 @@
 
     private DataHolder argumentHolder;
 
-    public GetRecordsMessageBodyReader(Converter provider, CswSourceConfiguration configuration) {
+    public GetRecordsMessageBodyReader(Converter converter, CswSourceConfiguration configuration) {
         xstream = new XStream(new XppDriver());
         xstream.setClassLoader(this.getClass()
                 .getClassLoader());
-        GetRecordsResponseConverter converter = new GetRecordsResponseConverter(provider);
```
Comment (tbatie) [inline catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReader.java]: This is definitely different. Can you explain why this change is correct?

File: catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java [hunk 6]
```diff
@@ -95,7 +91,7 @@ public void testConfigurationArguments() throws Exception {
         reader.readFrom(CswRecordCollection.class, null, null, null, httpHeaders, is);
 
         // Verify the context arguments were set correctly
-        verify(mockProvider, times(3)).unmarshal(any(HierarchicalStreamReader.class),
+        verify(mockProvider, times(1)).unmarshal(any(HierarchicalStreamReader.class),
```
Comment (tbatie) [inline catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java]: Why was this three times before and why is it only once now?

File: catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java [hunk 7]
```diff
@@ -118,9 +114,14 @@ public void testConfigurationArguments() throws Exception {
 
     @Test
     public void testFullThread() throws Exception {
-        CswTransformProvider provider = new CswTransformProvider(null, mockInputManager);
-
-        when(mockInputManager.getTransformerBySchema(anyString())).thenReturn(new CswRecordConverter());
+        List<Metacard> inputMetacards = new ArrayList<>();
```
Comment (tbatie) [inline catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java]: Is there a reason you choose to mock out a new provider instead of using the CswTransformProvider as before?

File: catalog/spatial/csw/spatial-csw-source/pom.xml [hunk 8]
```diff
@@ -172,6 +177,7 @@
                         <!-- Don't Export Anything -->
                         <Export-Package/>
                         <Embed-Dependency>
+                            spatial-csw-source-common,
                             spatial-csw-common,
                             spatial-ogc-common,
                             spatial-csw-transformer,
```
Comment (tbatie) [inline catalog/spatial/csw/spatial-csw-source/pom.xml]: This is a bundle :(

File: catalog/spatial/csw/spatial-csw-source/pom.xml (hunk 3 above)
Comment (tbatie) [inline catalog/spatial/csw/spatial-csw-source/pom.xml]: It wouldn't be a bad idea to move the test class for this base functionality back into this module. You could just create a concrete subclass in the test folder and use it to test against.

Then, the modules testing the concrete subclasses would only have to test the additional/changed functionality. Otherwise, you're relying on one particular concrete implementation to give you your code coverage.

File: catalog/spatial/csw/spatial-csw-source/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 9]
```diff
@@ -101,7 +105,11 @@
             <property name="outputSchema" value=""/>
             <property name="queryTypeQName" value=""/>
             <property name="queryTypePrefix" value=""/>
-            <property name="cswTransformProvider" ref="cswTransformProvider"/>
```
Comment (tbatie) [inline catalog/spatial/csw/spatial-csw-source/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: You should delete the `cswTransformProvider` on line 29.

**Comment by tbatie** [comment]
Besides the small questions, LGTM! :+1:

File: catalog/spatial/csw/spatial-csw-source/pom.xml (hunk 8 above)
Comment (clockard) [inline catalog/spatial/csw/spatial-csw-source/pom.xml]: I saw that and cringed myself. I think that changing this dependency could be part of the next tickets refactoring.

File: catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java (hunk 6 above)
Comment (clockard) [inline catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java]: Before it was using an actual transformer which was not necessary for this unit test. By not using the actual transformer i was able to untangle some of the dependency mess. Instead I just used a mock to return one result

File: catalog/spatial/csw/spatial-csw-source-common/pom.xml (hunk 4 above)
Comment (clockard) [inline catalog/spatial/csw/spatial-csw-source-common/pom.xml]: I will change it to 1.4. This was a copy paste but I don't remember where from.

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReader.java (hunk 1 above)
Comment (clockard) [inline catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReader.java]: Oops. Removed

File: catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java (hunk 2 above)
Comment (clockard) [inline catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java]: Not sure but I removed it

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReader.java (hunk 5 above)
Comment (clockard) [inline catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReader.java]: We were passing in a converter for the sole purpose of creating yet another converter. This created unwanted dependency on spatial-csw-transformer (a bundle!) so I just changed what the argument being passed in represented. Now I create a GetRecordsResponsConverter in the blueprint and pass that in as the converter instead.

File: catalog/spatial/csw/spatial-csw-source/pom.xml (hunk 3 above)
Comment (clockard) [inline catalog/spatial/csw/spatial-csw-source/pom.xml]: So right now all the functionality is in the abstract class so I think it makes sense to have the test with it. I agree that the concrete implementations should just be able to test their new/additional functionality but right now there is none. To vina's question the coverage went to the spatial-csw-source-common and there is really nothing to test in the concrete implementation. All that to say I don't know that I want to move the test back into spacial-csw-source. If you still want a test in spatial-csw-source let me know and we can work something out.

File: catalog/spatial/csw/spatial-csw-source-common/pom.xml [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.codice.ddf.spatial</groupId>
+        <artifactId>csw</artifactId>
+        <version>2.9.0-SNAPSHOT</version>
+    </parent>
+    <artifactId>spatial-csw-source-common</artifactId>
+    <packaging>jar</packaging>
+    <name>DDF :: Spatial :: CSW :: Source :: Common</name>
+    <dependencies>
```
Comment (lessarderic) [inline catalog/spatial/csw/spatial-csw-source-common/pom.xml]: Nit: This pom contains gazillions of dependencies. Would you mind putting the DDF ones together, the third party ones together and the test ones last?

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java [hunk 11]
```diff
@@ -222,7 +221,7 @@
     protected String configurationPid;
 
     static {
-        try (InputStream properties = CswSource.class.getResourceAsStream(
+        try (InputStream properties = AbstractCswSource.class.getResourceAsStream(
```
Comment (lessarderic) [inline catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java]: Could use http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/io/Resources.html#getResource(java.lang.String) here instead.

File: catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java [hunk 12]
```diff
@@ -136,7 +137,7 @@ public void testFullThread() throws Exception {
         config.setThumbnailMapping(CswRecordMetacardType.CSW_REFERENCES);
```
Comment (lessarderic) [inline catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java]: Delete commented out lines above (133-134).

File: catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java [hunk 13]
```diff
@@ -151,53 +152,23 @@ public void testFullThread() throws Exception {
...
-        assertMetacard(mc, expectedValues);
-
+        assertThat(mc.getId(), equalTo("metacard1"));
+        assertThat(mc.getTitle(), equalTo("title1"));
     }
 
     // verifies UTF-8 encoding configured properly when XML includes foreign text with special characters
     @Test
     public void testGetMultipleMetacardsWithForeignText() throws Exception {
+        List<Metacard> inputMetacards = new ArrayList<>();
```
Comment (lessarderic) [inline catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java]: These lines are very similar to the ones in `testFullThread()`. Any way we can extract the common code?

File: catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java [hunk 14]
```diff
@@ -151,53 +152,23 @@ public void testFullThread() throws Exception {
         List<Metacard> metacards = cswRecords.getCswRecords();
 
         assertThat(metacards, not(nullValue()));
```
Comment (lessarderic) [inline catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java]: Should be able to replace the four lines with `assertThat(metacards, contains(metacard));`.

File: catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java [hunk 15]
```diff
@@ -214,11 +185,9 @@ public void testGetMultipleMetacardsWithForeignText() throws Exception {
                 is);
         List<Metacard> metacards = cswRecords.getCswRecords();
```
Comment (lessarderic) [inline catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java]: `assertThat(metacards, is(empty()));` should work here.

File: catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java [hunk 16]
```diff
@@ -230,8 +199,8 @@ public void testReadProductData() throws Exception {
         ByteArrayInputStream dataInputStream = new ByteArrayInputStream(data);
         MultivaluedMap<String, String> httpHeaders = new MultivaluedHashMap<>();
         httpHeaders.add(CswConstants.PRODUCT_RETRIEVAL_HTTP_HEADER, "TRUE");
-        httpHeaders.add(HttpHeaders.CONTENT_DISPOSITION, String.format("inline; filename=%s",
-                "ResourceName"));
+        httpHeaders.add(HttpHeaders.CONTENT_DISPOSITION,
+                String.format("inline; filename=%s", "ResourceName"));
```
Comment (lessarderic) [inline catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java]: Not sure why we use `String.format` here since `"ResourceName"` is constant.

File: catalog/spatial/csw/spatial-csw-source-common/src/test/resources/logback-test.xml [hunk 17]
```diff
@@ -0,0 +1,26 @@
+<!--
```
Comment (lessarderic) [inline catalog/spatial/csw/spatial-csw-source-common/src/test/resources/logback-test.xml]: Good idea! We should move this to a common test library that anybody can use in their Maven modules to suppress logging during unit tests. Talk to me or @oconnormi to know more :-)

File: catalog/spatial/csw/spatial-csw-source-common/src/test/resources/logback-test.xml [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+-->
+<configuration>
+
+  <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
+    <encoder>
+      <pattern>%d{HH:mm:ss.SSS} [%thread] %-5level %logger{36} - %msg%n</pattern>
+    </encoder>
+  </appender>
+
+  <root level="INFO">
```
Comment (lessarderic) [inline catalog/spatial/csw/spatial-csw-source-common/src/test/resources/logback-test.xml]: If the purpose is to reduce the log outputs during tests, maybe we should make this `ERROR` instead of `INFO`?

File: catalog/spatial/csw/spatial-csw-source/pom.xml [hunk 19]
```diff
@@ -124,7 +129,7 @@
         <dependency>
             <groupId>xmlunit</groupId>
             <artifactId>xmlunit</artifactId>
-            <version>1.6</version>
+            <version>1.4</version>
```
Comment (lessarderic) [inline catalog/spatial/csw/spatial-csw-source/pom.xml]: Downgrading the dependency?

File: catalog/spatial/csw/spatial-csw-source/pom.xml (hunk 3 above)
Comment (lessarderic) [inline catalog/spatial/csw/spatial-csw-source/pom.xml]: `0.00` coverage makes me sad :-(

**Comment by lessarderic** [comment]
The "new" `AbstractCswSource` class is pretty big and seems to cover a lot of ground. Are there any plans to break it up into a few smaller, more manageable classes?

File: catalog/spatial/csw/spatial-csw-source/pom.xml (hunk 8 above)
Comment (tbatie) [inline catalog/spatial/csw/spatial-csw-source/pom.xml]: Definitely not a bad idea!

**Comment by clockard** [comment]
@lessarderic there is plans to break up the AbstractCswSource into more manageable and extensible modules. This pr is just the first step. I think that several of your comments on this pr will be addressed in the follow on ticket.

File: catalog/spatial/csw/spatial-csw-source/pom.xml (hunk 19 above)
Comment (clockard) [inline catalog/spatial/csw/spatial-csw-source/pom.xml]: I was told that 1.4 was the version we are using everywhere else so I changed it to be consistent

File: catalog/spatial/csw/spatial-csw-source/pom.xml (hunk 3 above)
Comment (clockard) [inline catalog/spatial/csw/spatial-csw-source/pom.xml]: Added note

File: catalog/spatial/csw/spatial-csw-source-common/pom.xml (hunk 10 above)
Comment (clockard) [inline catalog/spatial/csw/spatial-csw-source-common/pom.xml]: Reorganized the dependencies.

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java (hunk 11 above)
Comment (clockard) [inline catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java]: This kind of refactoring will be done in the follow on ticket

File: catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java (hunk 12 above)
Comment (clockard) [inline catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java]: Removed

File: catalog/spatial/csw/spatial-csw-source-common/src/test/resources/logback-test.xml (hunk 17 above)
Comment (clockard) [inline catalog/spatial/csw/spatial-csw-source-common/src/test/resources/logback-test.xml]: Might be something we consider for the follow on ticket or just a separate ticket all together. This is a carry over from the original CswSource testing.

File: catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java (hunk 14 above)
Comment (clockard) [inline catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java]: Fixed

File: catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java (hunk 13 above)
Comment (clockard) [inline catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java]: They are similar but I don't know if it is worth extracting for the two places we are using it since it is only a few lines.

File: catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java (hunk 16 above)
Comment (clockard) [inline catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/TestGetRecordsMessageBodyReader.java]: Fixed

File: catalog/spatial/csw/spatial-csw-source-common/src/test/resources/logback-test.xml (hunk 18 above)
Comment (clockard) [inline catalog/spatial/csw/spatial-csw-source-common/src/test/resources/logback-test.xml]: Again, carry over from the original CswSource tests so I am not sure what the intent was.

**Comment by tbatie** [comment]
:+1:

**Comment by tbatie** [comment]
Hero'd successfully
